File: src/283_Move_Zeroes/283_Move_Zeroes.c
#include <stdio.h>

void moveZeroes(int* nums, int numsSize);

int main(void) {
  int nums1[] = {0, 1, 0, 3, 12};
  moveZeroes(nums1, 5);

  for (int i = 0; i < 5; i++) {
    printf("%d ", nums1[i]);
  }
  printf("\n");

  int nums2[] = {0};
  moveZeroes(nums2, 1);
  printf("%d\n", nums2[0]);

  int nums3[] = {4, 2, 4, 0, 0, 3, 0, 5, 1, 0};
  moveZeroes(nums3, 10);
  for (int i = 0; i < 10; i++) {
    printf("%d ", nums3[i]);
  }
  printf("\n");
  return 0;
}

void moveZeroes(int* nums, int numsSize) {
  int left = 0, right = 1;

  while (left < numsSize && right < numsSize) {
    if (!nums[left] && !nums[right]) {
      right++;
      continue;
    }

    if (!nums[left]) {
      int temp = nums[left];
      nums[left] = nums[right];
      nums[right] = temp;
    }

    left++, right++;
  }
}

File: src/283_Move_Zeroes/283_Move_Zeroes.py
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        
        left, right = 0, 1
        numsLen = len(nums)

        while(left < numsLen and right < numsLen):
            while(left < numsLen and nums[left]): left += 1
            
            right = left + 1

            while(right < numsLen and not nums[right]): right += 1

            if(left < numsLen and right < numsLen):
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right += 1
            
        return nums
    

if __name__ == "__main__":
    sol = Solution()
    print(sol.moveZeroes([0,1,0,3,12]))
    print(sol.moveZeroes([0]))
    print(sol.moveZeroes([0,1]))
    print(sol.moveZeroes([4,2]))